File: Lib/Neon.Cadence/Internal/ClientMessages/NewWorkerReply.cs
//-----------------------------------------------------------------------------
// FILE:	    NewWorkerReply.cs
// CONTRIBUTOR: Jeff Lill
// COPYRIGHT:	Copyright (c) 2016-2019 by neonFORGE, LLC.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Diagnostics.Contracts;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Reflection;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Hosting.Server.Features;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Server.Kestrel.Core;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

using Neon.Cadence;
using Neon.Common;
using Neon.Diagnostics;
using Neon.IO;
using Neon.Net;
using Neon.Tasks;

namespace Neon.Cadence.Internal
{
    /// <summary>
    /// <b>proxy --> library:</b> Answers a <see cref="NewWorkerRequest"/>.
    /// </summary>
    [ProxyMessage(MessageTypes.NewWorkerReply)]
    internal class NewWorkerReply : ProxyReply
    {
        /// <summary>
        /// Default constructor.
        /// </summary>
        public NewWorkerReply()
        {
            Type = MessageTypes.NewWorkerReply;
        }

        /// <summary>
        /// The ID of the new worker.
        /// </summary>
        public long WorkerId
        {
            get => GetLongProperty("WorkerId");
            set => SetLongProperty("WorkerId", value);
        }

        /// <inheritdoc/>
        internal override ProxyMessage Clone()
        {
            var clone = new NewWorkerReply();

            CopyTo(clone);

            return clone;
        }

        /// <inheritdoc/>
        protected override void CopyTo(ProxyMessage target)
        {
            base.CopyTo(target);

            var typedTarget = (NewWorkerReply)target;

            typedTarget.WorkerId = this.WorkerId;
        }
    }
}

File: Go/src/temporal-proxy/internal/messages/new_worker_reply.go
//-----------------------------------------------------------------------------
// FILE:		new_worker_reply.go
// CONTRIBUTOR: John C Burns
// COPYRIGHT:	Copyright (c) 2016-2019 by neonFORGE, LLC.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package messages

import (
	internal "temporal-proxy/internal"
)

type (

	// NewWorkerReply is a ProxyReply of MessageType
	// NewWorkerReply.  It holds a reference to a ProxyReply in memory
	NewWorkerReply struct {
		*ProxyReply
	}
)

// NewNewWorkerReply is the default constructor for
// a NewWorkerReply
//
// returns *NewWorkerReply -> a pointer to a newly initialized
// NewWorkerReply in memory
func NewNewWorkerReply() *NewWorkerReply {
	reply := new(NewWorkerReply)
	reply.ProxyReply = NewProxyReply()
	reply.SetType(internal.NewWorkerReply)

	return reply
}

// -------------------------------------------------------------------------
// IProxyMessage interface methods for implementing the IProxyMessage interface

// Build inherits docs from ProxyReply.Build()
func (reply *NewWorkerReply) Build(e error, result ...interface{}) {
	reply.ProxyReply.Build(e)
	if len(result) > 0 {
		if v, ok := result[0].(int64); ok {
			reply.SetWorkerID(v)
		}
	}
}

// Clone inherits docs from ProxyReply.Clone()
func (reply *NewWorkerReply) Clone() IProxyMessage {
	newWorkerReply := NewNewWorkerReply()
	var messageClone IProxyMessage = newWorkerReply
	reply.CopyTo(messageClone)

	return messageClone
}

// CopyTo inherits docs from ProxyReply.CopyTo()
func (reply *NewWorkerReply) CopyTo(target IProxyMessage) {
	reply.ProxyReply.CopyTo(target)
}
